modules/tinymce/src/core/main/ts/api/util/Tools.ts

```diff
@@ -8,22 +8,23 @@
 import * as ArrUtils from '../../util/ArrUtils';
 import Env from '../Env';
 
-type ArrayCallback<T, R> = (x: T, i: number, xs: ReadonlyArray<T>) => R;
-type ObjCallback<T, R> = (value: T[keyof T], key: string, obj: T) => R;
+type ArrayCallback<T, R> = ArrUtils.ArrayCallback<T, R>;
+type ObjCallback<T, R> = ArrUtils.ObjCallback<T, R>;
 
 interface Tools {
   is (obj: any, type: string): boolean;
   isArray <T>(arr: any): arr is Array<T>;
   inArray <T>(arr: ArrayLike<T>, value: T): number;
-  grep <T>(arr: ArrayLike<T>, pred?: ArrayCallback<T, boolean>);
+  grep <T>(arr: ArrayLike<T> | null | undefined, pred?: ArrayCallback<T, boolean>): T[];
+  grep <T>(arr: Record<string, T> | null | undefined, pred?: ObjCallback<T, boolean>): T[];
   trim (str: string): string;
   toArray <T>(obj: ArrayLike<T>): T[];
   hasOwn (obj: any, name: string): boolean;
   makeMap <T>(items: ArrayLike<T> | string, delim?: string | RegExp, map?: Record<string, T | string>): Record<string, T | string>;
-  each <T>(arr: ArrayLike<T>, cb: ArrayCallback<T, any>, scope?: any): void;
-  each <T>(obj: T, cb: ObjCallback<T, any>, scope?: any): void;
-  map <T, U>(arr: ArrayLike<T>, cb: ArrayCallback<T, U>, scope?: any): Array<U>;
-  map <T, U>(obj: T, cb: ObjCallback<T, U>, scope?: any): Array<U>;
+  each <T>(arr: ArrayLike<T> | null | undefined, cb: ArrayCallback<T, void | boolean>, scope?: any): boolean;
```

Comment:
Should the `ArrayCallback<T, void>` and `ArrayCallback<T, boolean>` cases be broken up into different overloads so that they can have different return types? Or do we somehow come up with a boolean to return when the callbacks don't return anything?

Reply:
The `boolean` in the callback return type is just a flag to stop iterating. At all times this will return a boolean.